tests/test_fastavro.py

```diff
@@ -817,3 +817,52 @@ def test_cython_python():
         # CPython should use Cython.
         assert getattr(_reader, 'CYTHON_MODULE')
         assert getattr(_writer, 'CYTHON_MODULE')
+
+
+def test_union_records():
+    #
+    schema = {
+        'name': 'test_name',
+        'namespace': 'test',
+        'type': 'record',
+        'fields': [
+            {
+                'name': 'val',
+                'type': [
+                    {
+                        'name': 'a',
+                        'namespace': 'common',
+                        'type': 'record',
+                        'fields': [
+                            {'name': 'x', 'type': 'int'},
+                            {'name': 'y', 'type': 'int'},
+                        ],
+                    },
+                    {
+                        'name': 'b',
+                        'namespace': 'common',
+                        'type': 'record',
+                        'fields': [
+                            {'name': 'x', 'type': 'int'},
+                            {'name': 'y', 'type': 'int'},
+                            {'name': 'z', 'type': ['null', 'int']},
+                        ],
+                    }
+                ]
+            }
+        ]
+    }
+
+    data = {
+        'val': {
+            'x': 3,
+            'y': 4,
+            'z': 5,
+        }
+    }
+    encoded = bytearray()
```

Comment:
I think this will fail in `pypy`, because the pure Python implementation of `dump()` takes a file object rather than a `bytearray` parameter.

I introduced this inconsistency when implementing the Cython writer -- I did not realize that `dump()` is a public function of the library.

We / I need to think about how to address this.

Reply:
I think the simplest way to fix the library issue would be for `fastavro.dump()` to *always* call the pure Python version `fastavro._writer_py`, since that function already writes to a file object. It will be slower than the Cython version, which may not matter much, depending on how heavily the function is used.

A more proper fix would be for the Cython code to provide two functions, one writing to a file object and the other to a `bytearray`, and have `fastavro.dump()` use the former.

@scottbelden, @tebeka: Can you confirm that `fastavro.dump()` is intended to be used by application code to write single records to file objects? I'd like your thoughts on the above as well. I can create a PR for this if you want; I think it's a pretty simple change (< 1 hour).

Reply:
I have a PR up which should address this: https://github.com/tebeka/fastavro/pull/133.

Once it is merged, the `bytearray` in this code should be replaced with a file object. With that change, it should now work in both CPython and Pypy.